[PropertyWrappers] Fix a crash when using multiple wrappers of the same type

theblixguy opened the pull request:
At the moment, using multiple wrappers of the same type (consecutively) causes a crash in SILGen because `findOriginalPropertyWrapperInitialValue` does not consider that possibility. Instead, we end up accidentally setting the innermost call expression as the initial value and causing a crash.

```swift
@propertyWrapper
struct Foo<T> {
  init(wrappedValue: T) {
    self.wrappedValue = wrappedValue
  }

  var wrappedValue: T
}

@propertyWrapper
struct Bar<T> {
  init(wrappedValue: T) {
    self.wrappedValue = wrappedValue
  }

  var wrappedValue: T
}

struct PropertyContainer {
  @Foo @Bar var x: Int = 0 // This works
  @Foo @Foo var y: Int = 1 // This crashes
}
```

Resolves SR-10937.

theblixguy commented:
@swift-ci please smoke test

theblixguy commented:
@swift-ci please smoke test

theblixguy commented:
@swift-ci please smoke test

theblixguy commented:
@swift-ci please smoke test

DougGregor commented on include/swift/AST/Attr.h:
```diff
@@ -1501,6 +1502,9 @@ class CustomAttr final : public DeclAttribute,
   void setSemanticInit(Expr *expr) { semanticInit = expr; }
 
+  Expr *getInitialValueInit() const { return initialValueInit; }
+  void setInitialValueInit(Expr *expr) { initialValueInit = expr; }
```
If we're going to keep this new bit of state, I would prefer that it not be on `CustomAttr`. Rather, it should be part of the semantic information for the specific property wrapper (`PropertyWrapperBackingPropertyInfo`).

DougGregor commented on lib/Sema/TypeCheckPropertyWrapper.cpp:
```diff
@@ -677,6 +678,10 @@ Expr *swift::buildPropertyWrapperInitialValueCall(
          nullptr, /*implicit=*/true);
+
+      if (!lastWrapper->getInitialValueInit()) {
+        lastWrapper->setInitialValueInit(initializer);
```
The `initializer` is effectively going to get rewritten by `CSApply.cpp`, so this won't necessarily work.

DougGregor reviewed (commented):
Perhaps the existing approach isn't really wrong, but we could/should be checking the call depth to find the right initializer?

theblixguy committed:
[PropertyWrappers] When finding the initial value for the property, properly handle the situation where we have multiple attached property wrappers

If a property has multiple property wrappers attached, we'll have multiple nested calls, where each call's argument is a call to construct the next wrapper in the chain. However, when we use multiple wrappers consecutively, we cannot just rely on the call's type matching the innermost wrapper's type, because it will match the first wrapper in the sequence of consective wrappers and we'll end up crashing in SILGen. So, we should check if the call's argument is another call and look into that before checking the types.

theblixguy commented:
@DougGregor I guess so. We know how many wrappers we have and I think we should have the same number of `CallExpr`s as there are wrappers. I am not sure what the best way is to do that though (without a walker).

My initial approach was to check if the call's argument is another call and then look into that, before  we check the call's type against the innermost wrapper's type. However, I discarded that in favour of this approach to eliminate the walker. But seems like there are issues, so I have gone back to what I was doing before. 

Let me know if you think it's okay.

DougGregor reviewed (approved):
Yeah, I think I like this approach best.

theblixguy commented:
@swift-ci please smoke test
